### PR TITLE
Do reinitialization in serial if geom objects exist.

### LINKED ISSUE #5219: Sparsity augmentation for constraints does not happen in serial
Currently, FEProblem::reinitBecauseOfGhosting only reinitializes the jacobian for parallel runs. Thus, for serial runs, the sparsity is never augmented to add nonzero placeholders for constraint pairings.

The fix should be to reinit because of ghosting and/or because there are new geometric objects.

### PR DESCRIPTION
**Pull request by shartibartfast**
Closes #5219. Serial runs were not getting reinitialized jacobians
because there was no ghosting.

### CODE REVIEW

**Comment by shanestafford** [comment]
Marmot is failing in unrelated code.

**Comment by permcody** [comment]
@dschwen is working on a fix with his integration branch. We can invalidate that target when Marmot is repaired.

**Comment by friedmud** [comment]
@permcody What's the situation here?  This code shouldn't be causing any problems.  Do we just need to invalidate the app_test results?

**Comment by permcody** [comment]
This is part of the incompatible API change that Daniel pushed through  last week. An invalidate should take care of it.

File: framework/src/base/FEProblem.C [hunk 1]
```diff
@@ -2587,13 +2587,16 @@ FEProblem::computeUserObjects(ExecFlagType type/* = EXEC_TIMESTEP_END*/, UserObj
...
+FEProblem::reinitBecauseOfGhostingOrNewGeomObjects()
 {
-  // Need to see if _any_ processor has ghosted elems
-  dof_id_type ghosted = _ghosted_elems.size();
-  _communicator.sum(ghosted);
+  // Need to see if _any_ processor has ghosted elems or geometry objects.
+  bool needs_reinit = ! _ghosted_elems.empty();
+  needs_reinit = needs_reinit || ! _geometric_search_data._nearest_node_locators.empty();
+  needs_reinit = needs_reinit || ( _displaced_problem && ! _displaced_problem->geomSearchData()._nearest_node_locators.empty() );
+  dof_id_type do_reinit = static_cast<dof_id_type>(needs_reinit);
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: @shanestafford, not that it's a big deal but why cast this Boolean to an integer type only to use it as a Boolean again?

Comment (shanestafford) [inline framework/src/base/FEProblem.C]: My mistake; I merged two versions and forgot to clean up after myself. The
communicator.sum works fine with the bool.

**Comment by permcody** [comment]
@shanestafford - Are you planning to fix? I can take care of it if you are unable.
Thanks

**Comment by shanestafford** [comment]
I won't be able to get to it until next week, so yes, if you don't mind
fixing it. Thanks!

**Comment by permcody** [comment]
Fixing this in another PR - #5304

**Pull request closed without merging**